library/include/rocwmma/internal/flow_control.hpp

```diff
@@ -0,0 +1,136 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef ROCWMMA_FLOW_CONTROL_HPP
+#define ROCWMMA_FLOW_CONTROL_HPP
+
+namespace rocwmma
+{
+
+    namespace detail
+    {
+        // Perform synchronization across fragments(wavefronts) in a workgroup
+        struct amdgcn_barrier
+        {
+            __device__ static inline auto exec()
+            {
+                return __builtin_amdgcn_s_barrier();
+            }
+        };
+
+        template <int32_t mask = 0>
+        struct amdgcn_sched_barrier
+        {
+            enum : const int16_t
+            {
+                mask16 = mask
+            };
+
+            __device__ static inline auto exec()
+            {
+                return __builtin_amdgcn_sched_barrier(mask16);
+            }
+        };
+
+        template <int32_t priority = 0>
+        struct amdgcn_setprio
+        {
+            enum : const int16_t
+            {
+                priority16 = priority
+            };
+
+            __device__ static inline auto exec()
+            {
+                return __builtin_amdgcn_s_setprio(priority16);
+            }
+        };
+
+        template <int32_t vmcnt = -1>
+        struct amdgcn_s_waitcnt_vmcnt
+        {
+            enum : const int16_t
+            {
+                vmcnt16 = (vmcnt == -1) ? 0 : ((0xF) & vmcnt) & (((0x30) & vmcnt) << 10)
+            };
+
+            __device__ static inline auto exec()
+            {
+                return __builtin_amdgcn_s_waitcnt(vmcnt);
+            }
+        };
+
+        template <int32_t lgmcnt = -1>
+        struct amdgcn_s_waitcnt_lgkmcnt
+        {
+            enum : const int16_t
+            {
+                lgmcnt16 = (lgmcnt == -1) ? 0 : (((0xF) & lgmcnt) << 8)
+            };
+
+            __device__ static inline auto exec()
+            {
+                return __builtin_amdgcn_s_waitcnt(lgmcnt);
+            }
+        };
+
+        template <int32_t vmcnt = -1, int32_t lgmcnt = -1>
+        struct andgcn_s_waitcnt
+        {
+            enum : const int16_t
+            {
+                vmcnt16  = (vmcnt == -1) ? 0 : (((0xF) & vmcnt) & (((0x30) & vmcnt) << 10)),
+                lgmcnt16 = (lgmcnt == -1) ? 0 : (((0xF) & lgmcnt) << 8),
+                cnt      = vmcnt16 & lgmcnt16
+            };
+
+            __device__ static inline auto exec()
+            {
+                return __builtin_amdgcn_s_waitcnt(cnt);
+            }
+        };
+
```

Comment:
Instead of amdgcn_s_waitcnt_lgkmcnt and struct amdgcn_s_waitcnt_vmcnt defined again separately, use amdgcn_s_waitcnt as a base class:
```
template<int32_t vmcnt>
struct amdgcn_vmcnt : public amdgcn_s_waitcnt<vmcnt, -1> {};
template<int32_t lgkmcnt>
struct amdgcn_lgkmcnt : public amdgcn_s_waitcnt<-1, lgkmcnt> {};
```